callouts/java/service-callout/src/main/java/service/ServiceCallout.java

```diff
@@ -0,0 +1,286 @@
+    private String certKeyPath = "certs/pkcs8_key.pem";
+    private int serverThreadCount = 2;
+    private boolean enableInsecurePort = true;
+
+    /**
+     * Starts the service callout server.
+     * @throws IOException If an I/O error occurs while starting the server
+     */
+    public void start() throws IOException {
+
+        ServerBuilder<?> serverBuilder;
+        if (cert != null && certKey != null) {
+            logger.info("Secure server starting...");
+            serverBuilder = NettyServerBuilder.forPort(port)
+                    .sslContext(createSslContext(cert, certKey));
+        } else {
+            logger.info("Insecure server starting...");
+            serverBuilder = ServerBuilder.forPort(port);
+        }
+
+        server = serverBuilder.addService(new ExternalProcessorImpl()).build().start();
+        logger.info("Server started, listening on " + port);
+        Runtime.getRuntime()
+                .addShutdownHook(
+                        new Thread() {
+                            @Override
+                            public void run() {
+                                // Use stderr here since the logger may have been reset by its JVM shutdown hook.
+                                logger.info("*** shutting down gRPC server since JVM is shutting down");
+                                try {
+                                    ServiceCallout.this.stop();
+                                } catch (InterruptedException e) {
+                                    e.printStackTrace(System.err);
+                                }
+                                logger.info("*** server shut down");
+                            }
+                        });
+    }
+
+    /**
+     * Stops the service callout server.
+     * @throws InterruptedException If interrupted while waiting for the server to shut down
+     */
+    private void stop() throws InterruptedException {
+        if (server != null) {
+            server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Await termination on the main thread since the grpc library uses daemon threads.
+     * @throws InterruptedException If interrupted while waiting for the server to shut down
+     */
+    public void blockUntilShutdown() throws InterruptedException {
+        if (server != null) {
+            server.awaitTermination();
+        }
+    }
+
+    /**
+     * Processes a given request and returns the corresponding response.
+     * @param request The request to be processed
+     * @return The response to the processed request
+     */
+    public ProcessingResponse processRequest(ProcessingRequest request) {
+        ProcessingResponse.Builder builder = ProcessingResponse.newBuilder();
+
+        switch (request.getRequestCase()) {
+            case REQUEST_HEADERS:
+                Optional<HeadersOrImmediateResponse> response = onRequestHeaders(request.getRequestHeaders());
```

Comment:
I understand the fundamental issue with this code now, we need to pass the base builder class to the `on...` methods.

So we need to change all of the `onRequest` methods to also include the `builder.getImediateResponseBuilder()`
So that would be 

`onRequestHeaders(builder.getResponseHeadersBuilder(), builder.getImediateResponseBuilder(), request.getResponseHeaders());`